flask/helpers.py

```diff
@@ -401,14 +404,27 @@ def get_flashed_messages(with_categories=False, category_filter=[]):
     :param category_filter: whitelist of categories to limit return values
     """
     flashes = _request_ctx_stack.top.flashes
+    category_filter = set(category_filter)
     if flashes is None:
         _request_ctx_stack.top.flashes = flashes = session.pop('_flashes') \
-            if '_flashes' in session else []
+            if '_flashes' in session else {}
+    
     if category_filter:
-        flashes = list(filter(lambda f: f[0] in category_filter, flashes))
+        filtered_flashes = {}
+        for category in flashes:
+            if category in category_filter:
+                filtered_flashes[category] = flashes[category]
+    else:
+        filtered_flashes = flashes
+    
+    results = []
     if not with_categories:
-        return [x[1] for x in flashes]
-    return flashes
+        for flash_values in filtered_flashes.itervalues():
+            results += flash_values
+    else:
+        for flash_category, flash_values in filtered_flashes.iteritems():
+            results += map(lambda x: (flash_category, x), flash_values)
```

Comment:
NO! You should **never** use `map` with a lambda. It's inefficient and unpythonic. The proper construct here is a list comprehension:

    results += [(flash_category, x) for x in flash_values]